### PR TITLE
feat(servicecatalogappregistry): Initial L2 construct for Attribute Group

### PR DESCRIPTION
**Pull request by arcrank**
This PR creates the other main resource used by AppRegistry, the Attribute Group,
as an L2 construct.  We define the base construct here, associated methods will come
in forthcoming PR that create associations between these and applications.

Some special notes: the `attributes` of the attribute group are basically just a metadata
json like field.  We were able to define a JSII compatible type version that accepts the
variety of nested key-value pairs they can have.  There are no defined keys for it.

Testing done
------------------
* `yarn build && yarn lint && yarn test && yarn integ`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

Co-authored-by: Dillon Ponzo <dponzo18@gmail.com>

### CODE REVIEW

**Comment by arcrank** [comment]
One thing that will probably raise eyebrows is the attribute field.  We have explicit guidance to make this field required, even though an empty json is valid `{}`.  The reason for this is that we don't want end users to forget about this field when creating attribute group, since it's basically the main item of the resource.  It is possible to have workflows where defining it without attributes first is done.

Also this resource is very similar to the application construct created here  https://github.com/aws/aws-cdk/pull/15140 .  Will have a PR that adds methods to associate these coming out next, this PR just defines the base functionality.

File: packages/@aws-cdk/aws-servicecatalogappregistry/test/attribute-group.test.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    expect(stack).toHaveResourceLike('AWS::ServiceCatalogAppRegistry::AttributeGroup', {
+      Description: {
+        Ref: 'Description',
+      },
+    });
+  }),
+
+  test('attribute group created with a token attribute group name does not throw validation error', () => {
+    const tokenAttributeGroupName= new cdk.CfnParameter(stack, 'AttributeGroupName');
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/test/attribute-group.test.ts:101]:
```suggestion
    const tokenAttributeGroupName = new cdk.CfnParameter(stack, 'AttributeGroupName');
```

File: packages/@aws-cdk/aws-servicecatalogappregistry/test/integ.application.ts [hunk 2]
```diff
@@ -9,5 +9,4 @@ new appreg.Application(stack, 'TestApplication', {
   description: 'my application description',
 });
 
-app.synth();
-
+app.synth();
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/test/integ.application.ts:12]: Why this change? I think CDK files should end with a newline.

File: packages/@aws-cdk/aws-servicecatalogappregistry/test/attribute-group.test.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      });
+    }).toThrow(/Invalid attribute group name for resource/);
+  });
+
+  test('attribute group created with empty attributes', () => {
+    new appreg.AttributeGroup(stack, 'MyAttributeGroup', {
+      attributeGroupName: 'testAttributeGroup',
+      attributes: {},
+    });
+    expect(stack).toHaveResourceLike('AWS::ServiceCatalogAppRegistry::AttributeGroup', {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/test/attribute-group.test.ts:167]: An empty line before this line would be nice.

File: packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    const attributeGroup = new CfnAttributeGroup(this, 'Resource', {
+      name: props.attributeGroupName,
+      description: props.description,
+      attributes: props.attributes,
+    });
+
+    this.attributeGroupArn = attributeGroup.attrArn;
+    this.attributeGroupId = attributeGroup.attrId;
+  }
+  private validateAttributeGroupProps(props: AttributeGroupProps) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts:108]: Empty line before this line would be nice.

File: packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+   * Description for attribute group.
+   * @default - No description provided
+   */
+  readonly description?: string;
+
+  /**
+   * A JSON of nested key-value pairs that represent the attributes in the group.
+   * Attributes maybe an empty JSON '{}', but must be explicitly stated.
+   */
+  readonly attributes: Attribute;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts:54]: > One thing that will probably raise eyebrows is the attribute field. We have explicit guidance to make this field required, even though an empty json is valid `{}`. The reason for this is that we don't want end users to forget about this field when creating attribute group, since it's basically the main item of the resource. It is possible to have workflows where defining it without attributes first is done.

Not sure I follow, sorry. If it's valid to have this empty, then it should not be required.

File: packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+/**
+ * Attributes for the attribute group, defined by a nested list of
+ * key-value pairs.
+ */
+export interface Attribute {
+  [key: string]: string | number | boolean | Attribute |
+  string[] | number[] | boolean[] | Attribute[];
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts:33]: I don't think this adds much. I would remove this type, and declare `attributes` in `AttributeGroupProps `as

```ts
readonly attributes?: { [key: string]: any };
```

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great! Minor comments.

File: packages/@aws-cdk/aws-servicecatalogappregistry/test/integ.application.ts (hunk 2 above)
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalogappregistry/test/integ.application.ts:12]: added back

File: packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts (hunk 6 above)
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts:33]: Okay, my understanding that having `any` was frowned upon for typing.

File: packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts (hunk 5 above)
Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts:54]: Hence why I proactively tried to address this, since yes it does go against the convention of cdk.  I reached out to the dev team for comment and the gist again was basically that making the field optional would make it easier for customer to make a mistake of forgetting to populate an optional field when they would almost always want to.

Comment (arcrank) [reply packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts]: It is a rare but valid use case for it to be empty, the entire construct is basically a wrapper around this field.

Comment (skinny85) [reply packages/@aws-cdk/aws-servicecatalogappregistry/lib/attribute-group.ts]: OK. Let's leave it as-is then.

**Review by skinny85 — APPROVED** [review]
Looks great!
